### ISSUE TITLE
Deadlock in .net core consoleApp

### ISSUE THREAD

**Issue by galamayevg** [Azure/azure-storage-net#410]
There are test code, using c# azure storage client (last 8.1.0). If you run it as a .Net Framework ConsoleApp no problem, but if you run a .Net Core (netcoreapp1.0) ConsoleApp application sends first batch of messages and then just wait. How can I fix this?
```
class Program
    {
        static int parallelTasks = 100;
        static int count = 0;

        static void Main(string[] args)
        {
            MainAsync(args).GetAwaiter().GetResult();
        }

        static async Task MainAsync(string[] args)
        {
            var task1 = RunAsync("First task", "messages1", parallelTasks);
            var task2 = RunAsync("Second task", "messages2", parallelTasks);

            await Task.WhenAll(task1, task2);

            Console.WriteLine($"\r\n{task1.Result}\r\n{task2.Result}");
            Console.ReadLine();
        }

        static async Task<string> RunAsync(string name, string queueName, int parallelTasks)
        {
            var account = CloudStorageAccount.DevelopmentStorageAccount;
            var client = account.CreateCloudQueueClient();
            var queue = client.GetQueueReference(queueName);
            await queue.CreateAsync();

            var stopwatch = Stopwatch.StartNew();

            var tasks = new List<Task>();

            for (int i = 0; i < 1000; i++)
            {
                tasks.Add(AddMessageAsync(i, queue));

                while (tasks.Count > parallelTasks)
                {
                    var task = await Task.WhenAny(tasks);
                    tasks.Remove(task);
                }
            }

            await Task.WhenAll(tasks);

            var completed = stopwatch.ElapsedMilliseconds;
            stopwatch.Stop();
            await queue.DeleteAsync();
            return $"{completed} from {name}";
        }

        static async Task AddMessageAsync(int i, CloudQueue queue)
        {
            await queue.AddMessageAsync(new CloudQueueMessage(Encoding.UTF8.GetString(new byte[1024])));
            Interlocked.Increment(ref count);
            Console.Write($"\r{count}");
        }
    }
```

**Comment by erezvani1529** [comment]
Hi @galamayevg,

Thanks for sharing your issue. We had an issue with connections not being closed in AddMessage operation for RT/NetCore (in versions 8.0 and  8.0.1). However, the issue should be fixed with 8.1. I was not able to repro with your sample code either. Would you please double check that you have 8.1 installed and in case you have upgraded, retry with a clean build. 

Thanks,
Elham

**Comment by galamayevg** [comment]
Hi @erezvani1529,
I am create simple app for test https://github.com/galamayevg/NetCoreAzureStorage . I am check version is 8.1.

**Comment by erezvani1529** [comment]
I see. Thanks. We will investigate this and update with more findings.

**Comment by erezvani1529** [comment]
Hi @galamayevg,

Apologies for the delay. I tried to repro again with your code (StorageAccount/Emulator) but still can not see the issue.  One possible cause can be when there is a network error but the client never has its connection reset; it will continue waiting for the service to send data. You can set the MaximumExecutionTime option to set an upper bound on how long an operation can take (including retries).

Please let us know if setting MaximumExecutionTime helps and if not please share a process dump for further investigation.

Thanks,
Elham

**Comment by galamayevg** [comment]
Hi @erezvani1529,
[Dump 1](https://drive.google.com/open?id=0Bxem-V5BwvKRUXVpV2ExUWNWc00),
[Dump 2 (with MaximumExecutionTime = 5s)](https://drive.google.com/open?id=0Bxem-V5BwvKRUXVpV2ExUWNWc00).

Thanks.

**Comment by erezvani1529** [comment]
Hi @galamayevg ,

Thanks for sharing the dumps. Looks like you are running into thread-safety issues with the queue instances(as our objects are NOT guaranteed to be thread-safe). Please try using a new queue instance for each parallel task of adding the message to resolve your issue and let me know if the issue still persists.

Thanks,
Elham

**Comment by galamayevg** [comment]
Hi @erezvani1529,

I have updated library to the version 8.1.1 and target to .net core 1.1. For each request creates new instance CloudQueueClient and CloudQueue. Unfortunately only sent the first batch. I tested on azure and storage emulator. Here is the new link to the [dump](https://drive.google.com/file/d/0Bxem-V5BwvKRMmw0Q0FoUkRrMVE/view?usp=sharing) and a link to the [code](https://github.com/galamayevg/NetCoreAzureStorage). If after locking to wait for about 90 seconds, sending it continues but then stops again.

Thanks.

**Comment by erezvani1529** [comment]
Hi @galamayevg,

An update here that we are investigating the problem you have reported engaging folks from NetCore/CoreFx as well. Also to verify our repro, I am interested to know after how long do you wait before taking the dump? Does it hang infinitely or can you see that the number of threads are going up or some requests are being made(through fiddler for instance), or nothing at all? 

As a temporary workaround please try forcing the min worker threads to some higher number such as 300 following the instructions posted [here](https://github.com/dotnet/cli/blob/rel/1.0.0/Documentation/specs/runtime-configuration-file.md).

set COMPlus_ThreadPool_ForceMinWorkerThreads=300

Please let me know if the workaround helps.

Thanks,
Elham

**Comment by galamayevg** [comment]
Hi @erezvani1529,

System.Threading.ThreadPool.MinThreads: 300, works for me and also works
set COMPlus_ThreadPool_ForceMinWorkerThreads=300.

Before taking the dump, I wait about 60 seconds. At this time, the number of threads does not increase. Sent messages are visible in the queue. After sending 100 messages in parallel, defrosting occurs approximately after 90-120 seconds, the process continues, but after reaching the sent messages in 4000-5000 it is again frozen for 90-120 seconds. If you send 200 messages, defrosting occurs after 600-900 seconds.

Thanks.

**Comment by erezvani1529** [comment]
@galamayevg,

Great! Please feel free to close this issue if your problem is resolved since we will continue tracking the root cause fix for this issue with [issue #431 ](https://github.com/Azure/azure-storage-net/issues/431). 

Thanks!
